### ISSUE TITLE
No GPU/TPU found, falling back to CPU

### ISSUE THREAD

**Issue by ma-sadeghi** [patrick-kidger/diffrax#80]
Here's the full warning that I get (I do have a GPU):

```python
>>> import diffrax
2022-03-24 16:30:19.350737: I external/org_tensorflow/tensorflow/compiler/xla/service/service.cc:170] XLA service 0x55795c0d4670 initialized for platform Interpreter (this does not guarantee that XLA will be used). Devices:
2022-03-24 16:30:19.350761: I external/org_tensorflow/tensorflow/compiler/xla/service/service.cc:178]   StreamExecutor device (0): Interpreter, <undefined>
2022-03-24 16:30:19.353414: I external/org_tensorflow/tensorflow/compiler/xla/pjrt/tfrt_cpu_pjrt_client.cc:169] TfrtCpuClient created.
2022-03-24 16:30:19.353886: I external/org_tensorflow/tensorflow/stream_executor/tpu/tpu_platform_interface.cc:74] No TPU platform found.
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
```

Edit: I installed `diffrax` from conda-forge.

**Comment by patrick-kidger** [comment]
This is probably a JAX issue rather than a Diffrax issue.

1. What happens if you do
```python
>>> import jax
>>> jax.numpy.array(1) + 1
```
i.e. an operation on a JAX array? Does a warning appear?

2. What is the result of `jax.devices()`?
3. What happens if you install both JAX and Diffrax in a fresh environment via pip, instead of conda? It's worth noting that the conda-forge versions of JAX and Diffrax are maintained by the community (@dhirschfeld) rather than the official maintainers so they're a use-at-your-own-risk kind of affair.

**Comment by dhirschfeld** [comment]
IIUC `diffrax` will use the GPU capabilities of the underlying `jax` library so it sounds like you've installed a CPU only `jax` version. It would be useful to know the output of:
```
mamba list jax
```
```
mamba list diffrax
```
...also:
```
mamba info
```

**Comment by dhirschfeld** [comment]
However, it seems `conda-forge` doesn't yet support a GPU enabled `jax`:
https://github.com/conda-forge/jaxlib-feedstock/issues/34

**Comment by ma-sadeghi** [comment]
@dhirschfeld @patrick-kidger Thank you guys for your quick reply! Indeed, it was a `jax` issue. I created a fresh environment, installed `jax` and `diffrax`, also had to install `cudatoolkit-dev` from `conda-forge`, and finally I was able to get it working on GPU.


```
mamba list jax

# Name                    Version                   Build  Channel
jax                       0.3.4              pyhd8ed1ab_0    conda-forge
jaxlib                    0.3.2+cuda11.cudnn82          pypi_0    pypi
```

```
mamba list diffrax

# Name                    Version                   Build  Channel
diffrax                   0.0.5              pyhd8ed1ab_0    conda-forge
```

```
                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (0.19.1) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


     active environment : sci
    active env location : /home/amin/mambaforge/envs/sci
            shell level : 2
       user config file : /home/amin/.condarc
 populated config files : /home/amin/mambaforge/.condarc
                          /home/amin/.condarc
          conda version : 4.11.0
    conda-build version : not installed
         python version : 3.9.6.final.0
       virtual packages : __cuda=11.4=0
                          __linux=5.13.0=0
                          __glibc=2.31=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /home/amin/mambaforge  (writable)
      conda av data dir : /home/amin/mambaforge/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/amin/mambaforge/pkgs
                          /home/amin/.conda/pkgs
       envs directories : /home/amin/mambaforge/envs
                          /home/amin/.conda/envs
               platform : linux-64
             user-agent : conda/4.11.0 requests/2.26.0 CPython/3.9.6 Linux/5.13.0-35-generic ubuntu/20.04.4 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

**Comment by Dwarf1999** [comment]
I want to know how to use the command to do this

**Comment by dhirschfeld** [comment]
I'm not sure how @ma-sadeghi exactly got it to work, in particular I'm not sure why `cudatoolkit-dev` was required over just `cudatoolkit`. If I were to do it I would probably do something like:
```
mamba install -c conda-forge cudatoolkit=<cuda version> jax diffrax
conda remove --force jaxlib
pip install --no-deps jaxlib
```

Untested, so YMMV.

**Comment by ma-sadeghi** [comment]
@dhirschfeld The reason I had to do `cudatoolkit-dev` was because `cudatoolkit` doesn't automatically expose commands like `nvcc`, etc.

Here's how I did it in a fresh env:

```bash
conda create -n testenv python=3.8
conda activate testenv
mamba install -c conda-forge cudatoolkit-dev diffrax
conda remove --force jax jaxlib
pip install --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_releases.html
```

Note: Installing `cudatoolkit-dev` might take a while (the installation is actually pretty fast, the "Executing transaction" at the end took a while for me).

**Comment by Dwarf1999** [comment]
Sorry, I want to ask another question. Maybe I'm not too professional. My computer prompts:command 'mamba' not found       and so on.
So I want to know how to get this Mamba?

**Comment by ma-sadeghi** [comment]
@Dwarf1999 Not at all :)

Here's what I highly recommend to do (forget my previous instructions):

```
conda update -n base conda
conda install -n base conda-libmamba-solver

conda create -n testenv python=3.8
conda activate testenv
conda install -c conda-forge cudatoolkit-dev diffrax
conda remove --force jax jaxlib
pip install --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_releases.html
```

Here's some explanations: the first part replaces conda's default dependency resolver with the super fast `libmamba` that was [recently](https://www.anaconda.com/blog/a-faster-conda-for-a-growing-community) introduced. The second part is basically like before except there's no `mamba`! `mamba` is a third-party package that offers the same functionality as `conda` except much faster. You can install it (no longer need to) using `conda install -c conda-forge mamba`. The people at Anaconda realized how slow `conda` was and so that's why they're in the process of replacing the default resolver with `libmamba`. I think in a future release, it'll become the default, but until then you need to run the first two lines to manually make it so. Let me know if it worked for you.

**Comment by patrick-kidger** [comment]
I'm going to close this issue as it's JAX-related rather than Diffrax-related, but feel free to keep discussing things afterwards.

**Comment by dhirschfeld** [comment]
Just noting here that a CUDA-enabled `jax` is now available from `conda-forge` (for CUDA 11.2+ IIUC)

* https://anaconda.org/conda-forge/jaxlib/files
